[RFC] Last char in statusline does not show

MichaHoffmann opened the pull request:
Started working on #8466

Commit:
add (failing) test for multibyte statusline issue

Commit:
charset.c: fix length comparison

MichaHoffmann reviewed (commented):
If `buf_p + l = buf_e`  we are at the last byte before the NUL byte in our buffer. The break there was to early i think.

oni-link commented on src/nvim/charset.c:
```diff
@@ -370,7 +370,8 @@ size_t transstr_buf(const char *const s, char *const buf, const size_t len)
     if (l > 1) {
-      if (buf_p + l >= buf_e) {
+      //ignore lengths that do not fit the buffer
+      if (buf_p + l > buf_e) {
```
Similar case in line 387 for non printable character. 

The loop in the else case further down could also be fixed. It iterates over all fields of array `pcc` even if it is not filled up (try unicode 0x9f as statusline character). 
```diff

diff --git a/src/nvim/charset.c b/src/nvim/charset.c
index ab20996df..193d192ac 100644
--- a/src/nvim/charset.c
+++ b/src/nvim/charset.c
@@ -380,10 +380,10 @@ size_t transstr_buf(const char *const s, char *const buf, const size_t len)
         memmove(buf_p, p, l);
         buf_p += l;
       } else {
-        for (size_t i = 0; i < ARRAY_SIZE(pcc); i++) {
+        for (size_t i = 0; i < ARRAY_SIZE(pcc) && pcc[i]; i++) {
           char hexbuf[11];
           const size_t hexlen = transchar_hex(hexbuf, pcc[i]);
-          if (buf_p + hexlen >= buf_e) {
+          if (buf_p + hexlen > buf_e) {
             break;
           }
           memmove(buf_p, hexbuf, hexlen);
````

MichaHoffmann replied:
I'm sorry but isnt 0x9f the 'ß' ( im not too sure though )? In any case, i will correct line 387 and add a testcase for some non-printable unicode!

oni-link replied:
You can input unicode with `<CTRL-v>u9f<space>` in insert mode or on the command line.

MichaHoffmann replied:
I incorporated your comments! u9f was indeed broken; adding the &pcc[i] check resolved that. 
I added a test; please have a look.

Commit:
charset.c: adressed review comments, functional/api/vim_spec.lua:added test for unprintable unicode point

oni-link commented on test/functional/api/vim_spec.lua:
```diff
@@ -1267,5 +1268,44 @@ describe('API', function()
+      ]])
+    end)
+  end)
+  describe('statusline', function()
```
Why not group the following tests with the previous `describe()` block?

MichaHoffmann replied:
To be honest, i was not sure where to put them! Do you have an idea where the most fitting place would be?

oni-link replied:
I would have grouped them with the previous `describe()` block, because of the same name `statusline`.

MichaHoffmann replied:
Oh, that was just a copy paste error, i will correct it.

oni-link commented on test/functional/api/vim_spec.lua:
```diff
@@ -1267,5 +1268,44 @@ describe('API', function()
+                          |
+      ]])
+    end)
+  end)
```
We could add a test with up to `MAX_MCO` composing characters.

MichaHoffmann replied:
Should this test work? it only renders one <9f> in the bottom at the moment
```    it('MAX_MCO unicode control points', function()                                  
      local screen = Screen.new(24, 4)                                               
      screen:attach()                                                                
      command('set laststatus=2')                                                    
      feed([[                                                                        
      :set statusline=<C-v>u9f                                                       
      <C-v>u9f                                                                       
      <C-v>u9f                                                                       
      <C-v>u9f                                                                       
      <C-v>u9f                                                                       
      <C-v>u9f<SPACE><CR>                                                            
      ]])                                                                            
      screen:expect([[                                                               
      ^                        |                                                     
      ~                       |                                                      
      <9f><9f><9f><9f><9f><9f>|                                                      
                              |                                                      
      ]])                                                                            
    end)                                                                             
```

oni-link replied:
Try putting all the character on one line: `:set statusline=<C-v>u9f<C-v>u9f<C-v>u9f<C-v>u9f<C-v>u9f<C-v>u9f<CR>`

Also `U+009f` is not a [composing character](https://en.wikipedia.org/wiki/Combining_character). To input unicode with more than two hex digits you can use `<C-v>U...`:
a + U+fe20: a︠
a + U+fe20 + U+fe21: a︠︡

MichaHoffmann replied:
This behaves weird: 

```
    it('MAX_MCO (6) unicode control points', function()                              
      local screen = Screen.new(24, 4)                                               
      screen:attach()                                                                
      command('set laststatus=2')                                                    
      feed(string.format(":set statusline=a%s%s%s%s%s%s<CR>",                          
      "<C-v>ufe20", "<C-v>ufe21", "<C-v>u20df", "<C-v>ufe26",                          
      "<C-v>ufe2e", "<C-v>u20de"))                                                     
      screen:expect([[                                                                 
      ^                        |                                                       
      ~                       |                                                        
      a︠︡⃟︦︮⃞                       |                                                        
                              |                                                        
      ]])                                                                              
    end)
```

results in 
```
Expected:
  |^                        |
  |~                       |
  |a︠︡⃟︦︮⃞                       |
  |*                        |
Actual:
  |^                        |
  |~                       |
  |a︠︡⃟︦︮⃞                       |
  |*:set statusline=a︠︡⃟︦︮⃞       |
```
although the same command works without composing characters. I'm not sure why that does not work ( the 6 composing chars are rendered correctly )

MichaHoffmann commented on src/nvim/charset.c:
```diff
@@ -370,7 +370,8 @@ size_t transstr_buf(const char *const s, char *const buf, const size_t len)
+      if (buf_p + l > buf_e) {
         break;
       }
       int pcc[MAX_MCO + 2];
```
Thats very true, i'll do that.

MichaHoffmann replied:
`utfc_ptr2char` says it wants a buffer with space for at least `MAX_MCO+1` chars, and since we put in `&pcc[1]` `pcc[MAX_MCO+1]` seems to be ok. Propably the fix would be 
```
if (i <= MAX_MCO)      /* last composing char must be 0 */                         
  pcc[i] = 0;  
``` 
in `utfc_ptr2char` so the 0 is set even if we found MAX_MCO composing chars.

MichaHoffmann replied:
also `utfc_ptr2char` stops at 
```
    for (;; ) {                                                                      
      pcc[i++] = cc;                                                                 
      if (i == MAX_MCO)                                                              
        break;   
```
so pcc only holds MAX_MCO-1 composing chars; or am i wrong?

MichaHoffmann replied:
I still dont understand: the loop in `utfc_ptr2char` breaks on `i == MAX_MCO`; if `pcc` is smaller then `MAX_MCO+1` we would write out of bounds, or am i wrong?

MichaHoffmann replied:
Yep you are right, i re-read the function more carefully this time. ( should i change the docstring in this PR? )

MichaHoffmann commented:
Is this expected behaviour ? 
Test:
```
      local screen = Screen.new(24, 4)                                           
      screen:attach()                                                            
      command('set laststatus=2')                                                
      feed(string.format(":set statusline=a%s%sb<CR>",                           
      "<C-v>ufe20", "<C-v>ufe21"))                                                   
      screen:expect([[                                                               
      ^                        |                                                     
      ~                       |                                                      
      a︠︡b                      |                                                      
                              |                                                      
      ]])   
```

Result:
```
Expected:
  |^                        |
  |~                       |
  |a︠︡b                      |
  |*                        |
Actual:
  |^                        |
  |~                       |
  |a︠︡b                      |
  |*:set statusline=a︠︡b      |
```

oni-link commented:
> Is this expected behaviour ?

Does it depend on the use of `command(...)` or `feed(...)`?

MichaHoffmann commented:
using command results in 
```Expected:
  |^                        |
  |~                       |
  |*a︠︡b                      |
  |                        |
Actual:
  |^                        |
  |~                       |
  |*a<C-v>ufe20<C-v>ufe21b  |
  |                        |
```

oni-link commented:
If I input the characters directly it works:
`command('set statusline=a︠︡b')` 

To see what code points form a character, move the cursor on the character and press `ga` or for the byte representation use `g8`.

MichaHoffmann commented:
Thanks, that worked ( and viewing the code points helps too, these continuation characters are getting crowded .. )

Commit:
charset.c: review comment and hexbuf; add test for MAX_MCO combination chars

oni-link commented on test/functional/api/vim_spec.lua:
```diff
@@ -1267,5 +1268,55 @@ describe('API', function()
+      screen:attach()
+      command('set laststatus=2')
+      command('set statusline=o̸⃯ᷰ⃐⃧⃝')
+      --o + U+1DF0 + U+20EF + U+0338 + U+20D0 + U+20E7 + UU+20DD
```
UU+20DD -> U+20DD

Also same test with leading non-printable character followed by six combining characters?

Tests could be written a bit shorter if we use something like this
```lua
    before_each(function()
      --clear()
      screen = Screen.new(40, 4)
      screen:attach()
      command('set laststatus=2')
    end)

    after_each(function()
        screen:detach()
    end)

    ...

    it('last char shows', function()
      command('set statusline=abc')
      screen:expect([[
      ^                                        |
      ~                                       |
      abc                                     |
                                              |
      ]])
    end)

```

MichaHoffmann replied:
Ok i've done that.

Commit:
vim_spec.lua test for non printable point followed MAX_MCO combination points

oni-link commented on test/functional/api/vim_spec.lua:
```diff
@@ -1267,5 +1268,68 @@ describe('API', function()
   end)
+  describe('statusline', function()
+    before_each(function()
+      screen = Screen.new(40, 4)
```
What `screen` is used here? Should it be local to the `describe()` block?

MichaHoffmann replied:
You are right, made it local to the describe scope.

oni-link commented on test/functional/api/vim_spec.lua:
```diff
@@ -1267,5 +1268,68 @@ describe('API', function()
+      screen = Screen.new(40, 4)
+      screen:attach()
+      command('set laststatus=2')
+		end)
```
For me this has not the right indentation because of tabs.

MichaHoffmann replied:
Sorry, dont know how that happend! Fixed it.

oni-link commented:
`make lint` complains about
```
src/nvim/charset.c:364:  Function attribute line should have 2-space indent  [whitespace/indent] [5]
src/nvim/charset.c:373:  Should have a space between // and comment  [whitespace/comments] [4]
src/nvim/charset.c:385:  At least two spaces is best between code and comments  [whitespace/comments] [2]
src/nvim/charset.c:385:  Should have a space between // and comment  [whitespace/comments] [4]
```

MichaHoffmann commented on test/functional/api/vim_spec.lua:
```diff
@@ -1267,5 +1268,68 @@ describe('API', function()
+    end)
+
+    it('last char shows (multibyte)', function()
+      command('set laststatus=2')
```
Thats very true, removed it in the testcases.

MichaHoffmann commented:
mh i dont get the make lint error because it stops after complaining about mbyte.c:1125!

Commit:
charset.c: fixed lint errors

Commit:
vim_spec.lua: cleaned up statusline tests

oni-link commented on src/nvim/charset.c:
```diff
@@ -370,20 +370,21 @@ size_t transstr_buf(const char *const s, char *const buf, const size_t len)
     if (l > 1) {
-      if (buf_p + l >= buf_e) {
+      // ignore lengths that do not fit the buffer
+      if (buf_p + l > buf_e) {
```
A check like this needs to be added to the `l==1` branch. `transchar_byte()` could return a string of length greater 1.

oni-link commented:
LGTM after adding a length check to the `l==1` branch in `transstr_buf()`.

bfredl commented on test/functional/api/vim_spec.lua:
```diff
@@ -1267,5 +1268,65 @@ describe('API', function()
       eq(expected, nvim("list_uis"))
     end)
   end)
+  describe('statusline', function()
```
these tests should be in `ui/`, maybe `ui/multibyte_spec.lua`

Commit:
charset.c: review comments; vim_spec.lua; multibyte_spec.lua: moved tests

MichaHoffmann commented:
Addressed your comments, moved the tests and added a check for length in the third branch. 
The statusline in the tests is arbitrary; should it go? 
There is some redundancy between transstr_buf and transstr_len and inside the conditionals inside those functions. Now that there are the tests one could think about refactorings.

justinmk commented on test/functional/ui/multibyte_spec.lua:
```diff
@@ -119,4 +121,66 @@ describe("multibyte rendering", function()
   end)
 end)
+  
+describe('statusline', function()
```
describe('multibyte rendering: statusline', function()

justinmk commented:
Merged, thank you @MichaHoffmann and @oni-link !
